### ISSUE TITLE
Changes to polite aria-live region are repeated too many times in Chrome and Firefox

### ISSUE THREAD

**Issue by alex-barstow** [nvaccess/nvda#7996]
### Steps to reproduce:
1. Open this [test page](http://solutions.brightcove.com/abarstow/BC-39639/nvda-test.html) in Chrome
2. Click the button
3. Listen for repetitive announcements 
4. Repeat 2 and 3 to toggle the button's state

### Expected behavior:
The updated text should be announced ~1-2 times (2 would be understandable since the button click updates both the`title` attribute of the button element and the `innerText` of the child <span> element)

### Actual behavior:
The updated text is repeated 6 times in Chrome (ex. "Text B, Text B, B, Text B, Text B, Text B") and 3 times in Firefox (ex. "Text A, Text A, Text A")

### System configuration:
NVDA version: v2017.4

NVDA Installed or portable:
Installed

Windows version:
Windows 10 Pro Version 1709 Build 16299.192

Browser versions:
Chrome 63 and 64
Firefox 58

### Other questions:

Does the issue still occur after restarting your PC?
Yes

Have you tried any other versions of NVDA?
> Yes, v2015.3 (same results, except there are 5 repetitions instead of 6 in Chrome)

**Comment by RichCaloggero** [comment]
## Having sam eissue

This only occurs with React; does not occur using direct dom manipulation. See more explanation and tests here:

https://github.com/RichCaloggero/react-test

**Comment by RichCaloggero** [comment]
## THere may be a fix for this

The following github repo contains a set of react components which seem to fix the repetition issue.

So far, only tested in Firefox.

URL: https://github.com/AlmeroSteyn/react-aria-live

**Comment by jcsteh** [comment]
@RichCaloggero: I think the issue you are experiencing is different to this one, even though it has the same symptoms. As discussed via email, yours relates to events being fired for each area of the text that changed, and there wouldn't be "repetition" if aria-atomic wasn't set.

In this test case, the three announcements occur as follows:

1. The button is focused and its name (as set by the title attribute) changes. Whether or not it's a live region, we would report the change in that case. While I understand this is a distilled test case, there's no good "real world" reason to make a focused button a live region, as its name will always be read when it is changed while it is focused.
2. The text of the button changes, so a text change event gets fired. NVDA's live region code reports the change.
3. The name of the button changes (see 1), so a name change event is fired. Because live regions can include graphics, etc. which have names (e.g. alt attribute) that aren't part of the "text", NVDA's live region code also handles name changes. Thus, the change gets reported.

Ideally, we shouldn't do 3) in this case.

@MichaelDCurran, my understanding is that we report name changes for live regions because of alt text (#3329). Could we restrict this check so we only do this if the element in question has explicit-name: true? For that matter, should we also avoid retrieving text if explicit-name is true?

I'm not sure why there are even more repeats in Chrome, but this may possibly improve with #8043.

**Comment by RichCaloggero** [comment]
## Fix with this component

Install `react-aria-live` from npm.  See it's github page here:

https://github.com/AlmeroSteyn/react-aria-live

### How it works

Basically, react changes the dom as efficiently as possible. This means, among other things, that if you have some text in a live region and you then update it (i.e. cause the component to update), react will not simply rewrite the entire thing with textContent or appendChild; it will change the actual text nodes. For instance, if you have this:

```
<div aria-live="polite" aria-atomic="true">
{message}
</div>
```

and you change `message`, then text text nodes inside the div will be updated, not removed and rewritten.

The browser will attempt to do this by changing character ranges. For instance if previous contents were "this is a test", and new contents are "this was a simple test", the browser would change three ranges and then fire three events which the accessibility API passes to the screen reader.

The fix then is to empty the container before writing new text to it, but only do this when the new text is different. A bit tricky in react without directly manipulating the dom; use two live regions and alternate between them.

Here is a very simple component that does this, but highly suggest using the `react-aria-live` component mentioned above as it covers more corner cases and is well written and documented.

```
class Status extends React.Component {
constructor (props) {
super (props);
this.state = {alternate: false};
} // constructor

static getDerivedStateFromProps(props, state) {
return {alternate: !state.alternate};
} // getDerivedStateFromProps

render () {
  const {alternate} = this.state;
  const {message, ..._props} = this.props;
  
  return (
  <div>
    <div {..._props}>
  {alternate? message : null}
  </div>
      <div {..._props}>
  {!alternate? message : null}
  </div>
  </div>
  );
} // render
}; // Status
```

**Comment by Adriani90** [comment]
Small progress in Chrome 74 with NVDA 2019.1.1, the test is repeated only 4 times instead of 6 👍 :-)

**Comment by Yegorich555** [comment]
One more issue. If a child contains a couple items and you change aria-hidden="true" to aria-hidden="false" it will announce as much as items in the child, but expected only once!!!

**Comment by Adriani90** [comment]
@Yegorich555 I guess the slide 2 is still visible although aria-hidden=true and this leads to the objects being repeated. Right? Usually aria-hidden=true is used together with display:none or visibility:hidden like in this example:
https://www.w3.org/WAI/PF/aria-1.1/states_and_properties#aria-hidden
Using only aria-hidden=true would suffice for screen readers which access information only through the DOM directly. But as far as I understand NVDA is not using only the DOM but also other accessibility platforms.

But I think @jcsteh could explain much better than me :-)

**Comment by Yegorich555** [comment]
@Adriani90 you right. Slide 2 is still visible for a couple seconds, but aria-hidden='true' must fix twice-reading... One more detail: aria-relevant='text' fixes twice-reading but it doesn't work when we have 'div' with nested 'p'. So the following example works for me
```html
<div tabindex='0' aria-live='polite' aria-relevant='text'>
    <div id='slide-1' aria-hidden="false">
       <img src='...' alt='Banner 1'/>
       <div class="visuallyhidden"> Some text here and here</div>
       <div aria-hidden="true">
           <p> Some text here</p>
           <p> and here</p>
       </div>
    <div>
   <div id='slide-2' aria-hidden="true">
       <img src='...' alt='Banner 2'/>
       <div class="visuallyhidden"> Some text here and here etc.</div>
       <div aria-hidden="true">
           <p> Some text here</p>
           <p> and here etc.</p>
       </div>
    <div>
</div>
```

**Comment by AleksandrHovhannisyan** [comment]
This is a bug even without React. Simple demo here: https://gui-challenges.web.app/theme-switch/dist/. The button uses `aria-live="polite"`, but changes are narrated three times in succession.

**Comment by RichCaloggero** [comment]
1. I hear the announcement exactly twice with Firefox and latest NVDA (2021.3.1)
2. my suspician is that when the label changes, the live region triggers and new text is read, but because the button also has focus it is re-read
This case should never arrise; why would you put live region marup on a button?  If you want to communicate the state of a button, make it a toggle button, or simply change the label as you are doing and NVDA will read the new one (i.e. drop the live region markup and the announcement will only happen once).

**Comment by seanbudd** [comment]
@AleksandrHovhannisyan - is this being tracked by the React bug tracker? Could you link it if so?

**Comment by AleksandrHovhannisyan** [comment]
@seanbudd Not sure—where would I find that bug tracker? Unless I misunderstood you: I don't think this is a React bug.

**Comment by seanbudd** [comment]
@AleksandrHovhannisyan - apologies I misread your comment

**Comment by Adriani90** [comment]
The use case in https://github.com/nvaccess/nvda/issues/7996#issuecomment-1018515212 is still reproducible in NVDA 2024.4 with Firefox 131 (NVDA repeats twice) and Chromium Canary 132 (NVDA repeats 3 times).

**Comment by Made-of-Clay** [comment]
To enter and contribute to the conversation, I've been able to reproduce this behavior with something as simple as opening the dev tools, adding `aria-live="polite"` to a container, then appending content like `element.innerHTML += '<label>Foo<input></label>'`. The text "Foo" will be read twice.

**Comment by WofWca** [comment]
I'd guess this happens because first it reads the label of the element that has changed, assuming that it's only its contents that changed, but then it also notices that the label itself has changed, and reads the new label.

**Comment by oooasis** [comment]
**Tracker**: nvaccess/nvda — append to existing open issue #7996 (do not open a new issue)
**Contract**: C-17 — UNCLEAR (no normative text on attribute-only changes)
**Environment**: Windows 10 19045; NVDA 2026.1.1 (eSpeak; NVDA's UIA-for-Chromium setting = No); Chrome 149.0.7827.54 (CDP-driven); Firefox 155.0.1 (WebDriver BiDi-driven). Data: probe round-2, 2026-09-06, run ids chrome-20260906T123244Z and firefox-20260906T125205Z, 10 fresh navigations per engine per fixture; a formal rerun under the frozen protocol is scheduled for 2026-09-15/16.

## Summary
Only aria-label of a pre-existing polite live region changes (text content unchanged). NVDA announces the change on both engines; with Chrome it is announced twice per change, with Firefox once. We cannot determine from outside whether the duplication comes from duplicated browser events or from NVDA's processing; please redirect if a browser tracker is the right place.

## Related entries (searched 2026-09-07)
nvaccess/nvda #7996 (open, 2018): "Changes to polite aria-live region are repeated too many times in Chrome and Firefox when title/label of control is changed" — this report is additional data for that issue.

## Expected
No normative text (Core-AAM defines no text-change event for aria-label changes on a live region).

## Observed
Chrome 149: announced in 10/10 trials, 2 announcements per change (mode 2). Firefox 155: 10/10, 1 announcement per change (mode 1).


## Minimal fixture
Open the page, press the button (`#go`); the message contains a per-run nonce (`?nonce=` query, fallback literal).

```html
<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8">
<title>PercA11y probe c2-004-l0v0p0a0r0b0u4c0f0n0k0</title>
<meta name="probe-vector" content='{"lifecycle":"pre_existing_empty","visibility":"none","politeness":"polite","atomic":"default","relevant":"default","busy":"none","update_kind":"attribute_only","cadence":"single","focus":"none","container":"light_dom","channel":"live_region"}'>
<meta name="probe-expectation" content="RECORD_ONLY">
<meta name="probe-allowed" content='{"presence":"RECORD_ONLY","count_min":0,"count_max":1,"text_mode":"any_issued"}'>
</head>
<body data-probe="c2-004-l0v0p0a0r0b0u4c0f0n0k0">
<main>
<h1>Probe program c2-004-l0v0p0a0r0b0u4c0f0n0k0</h1>
<h2 id="editor" tabindex="-1">Editor</h2>
<button id="go">Run</button>
<div id="mount"></div>
<script>
var nonce=new URLSearchParams(location.search).get('nonce')||'NONCE0';
var T=function(i){return 'Item '+nonce+'-'+i;};
var OFF=[0],doc=document,root=null,reg=null;
function region(t,b){var e=doc.createElement('div');e.id='live';e.setAttribute('aria-live','polite');
var s=doc.createElement('span');s.textContent=t;e.appendChild(s);return e;}
function upd(i){var t=T(i);reg.setAttribute('aria-label',t);}
function seq(){OFF.forEach(function(d,i){var f=function(){upd(i+1);if(i===OFF.length-1)done();};if(d){setTimeout(f,d);}else{f();}});}
function done(){}
function trigger(){seq();}
function setup(){root=document.getElementById('mount');reg=region('',false);root.appendChild(reg);document.body.setAttribute('data-probe-ready','1');}
addEventListener('load',setup);
document.getElementById('go').addEventListener('click',trigger);
</script>
</main>
</body>
</html>
```

## Provenance
The generated-program vector is in `<meta name="probe-vector">`. Raw NVDA speech logs, timestamps and evaluation tooling are available on request.
